CONTRIBUTING.md

```diff
@@ -570,6 +570,58 @@ It's a "to do" item and should be replaced if possible. `Any` is used when
 it's not possible to accurately type an item using the current type system.
 It should be used sparingly.
 
+### "The `Any` trick"
+
+Consider the following signature of `re.Match[str].group`:
+
+```python
+class Match:
+  def group(self, __group: str | int) -> str | Any: ...
+```
+
+The `str | Any` seems unnecessary and weird at first.
+Because `Any` includes all strings, you would expect `str | Any` to be
+equivalent to `Any`, but it is not. To understand the difference,
+let's look at what happens when type-checking this simplified example:
+
+Suppose you have a legacy system that for historical reasons has two kinds
+of user IDs. Old IDs look like `"legacy_userid_123"` and new IDs look like
+`"456_username"`. The function below is supposed to extract the name
+`"USERNAME"` from a new ID, and return `None` if you give it a legacy ID.
+
+```python
+import re
+
+def parse_name_from_new_id(user_id: str) -> str | None:
+    match = re.fullmatch(r"\d+_(.*)", user_id)
+    if match is None:
+        return None
+    name_group = match.group(1)
+    return name_group.uper()  # This line is a typo (`uper` --> `upper`)
+```
+
+Regexes are often used so that the regex matches any string, or the string
+has already been validated so that the regex will match. But type checkers
+don't know this, so complaining about the `None` would get very annoying.
+
+* `-> Any` would mean "please do not complain" to type checkers.
```

Comment:
This probably needs something to bring the reader's attention back to the return type.

```suggestion
Let's now consider typeshed's `-> str | Any` annotation of the `.group()` method:

* `-> Any` would mean "please do not complain" to type checkers.
```